Tests: enforce critical class instance attributes in functional tests, fix segwit test specificity

JustinTArthur opened the pull request:
No extra attributes will be able to be added to instances of the C++ class ports or of other critical classes without causing an exception. Helps prevent adding or depending on attributes that aren't in the intended object structure. It may prevent issues such as the one fixed in bitcoin/bitcoin#14300.

This request fixes the erroneous version attribute used in the p2p_segwit.py functional tests. This pull includes the commit from bitcoin/bitcoin#14300.

practicalswift committed:
test: Fix broken segwit test

JustinTArthur committed:
Fix for incorrect version attr set on functional test segwit block.

JustinTArthur committed:
Strictly enforce instance attrs in critical functional test classes.

Additionally, removed redundant parentheses and added PEP-8 compliant
spacing around those classes.

JustinTArthur commented:
Travis tests should be re-run following the merge of #14300.

practicalswift commented:
What an excellent first-time contribution! Thanks!

This is my favourite contribution: it not only fixes current code but also fixes **future** code by introducing a mechanism that helps us avoid introducing new instances of the same class of issue. Exactly the type of contribution that I was referring to here: https://github.com/bitcoin/bitcoin/pull/14217#issuecomment-423812131:

> Personally I'm much more excited to see a PR submitted which strengthens the long term security of the project in some form compared to when I see a PR submitted which implements some shiny new feature :-)

utACK 965c8ff9c7539719eb36bd46297358e33b571766

MarcoFalke commented on test/functional/p2p_segwit.py:
```diff
@@ -771,8 +771,8 @@ def test_p2sh_witness(self):
         # segwit-aware would also reject this for failing CLEANSTACK.
         test_transaction_acceptance(self.nodes[0], self.test_node, spend_tx, with_witness=False, accepted=False)
```
Would be nice to demonstrate that this fails due to the correct reason: "non-mandatory-script-verify-flag (Witness program was passed an empty witness)"

You can achieve this with the `assert_debug_log` helper.

JustinTArthur replied:
I agree. Thanks for the tip on assert_debug_log. I'm more familiar with Python than I am of the test suite. Addressed in a new commit.

MarcoFalke commented on test/functional/p2p_segwit.py:
```diff
@@ -771,8 +771,8 @@ def test_p2sh_witness(self):
+        # Try to put the witness script in the scriptSig, should also fail.
+        spend_tx.vin[0].scriptSig = CScript([p2wsh_pubkey, b'a'])
         spend_tx.rehash()
         test_transaction_acceptance(self.nodes[0], self.test_node, spend_tx, with_witness=False, accepted=False)
```
Same here: The reject reason is just the same as above, whereas it now should be "mandatory-script-verify-flag-failed (Script evaluated without error but finished with a false/empty top stack element)"

JustinTArthur replied:
Addressed in a new commit.

MarcoFalke commented:
Concept ACK. I guess this could also be implemented with a "frozen" decorator, which freezes the attributes after `__init__`.

It worries me that the tests just pass with or without this fix, which doesn't add much value to the fix. Would you mind changing the tests itself, such that they would fail without, but pass with the change. (See inline comments)

JustinTArthur committed:
Check for specific tx acceptance failures based on script signature

JustinTArthur commented:
> Concept ACK. I guess this could also be implemented with a "frozen" decorator, which freezes the attributes after `__init__`.

I'm not opposed to moving to that at some point. Using `__slots__`  everywhere makes the code more verbose, where Python can otherwise excel at being concise. Unfortunately, these tests seem to do a bit of assignment after initialization so some refactoring would be necessary.

Added some checks for specific transaction acceptance failures in the debug log per @MarcoFalke's suggestion. Now at least the scriptSig typo would have caused a failure.

Thanks for the feedback, @MarcoFalke @gmaxwell and @practicalswift.

JustinTArthur commented:
@MarcoFalke I've created an alternate implementation that uses a freezing decorator as you've suggested:
https://github.com/bitcoin/bitcoin/compare/master...JustinTArthur:functional-test-attr-enforcement-decorator
Decorator implementation isn't great to look at, but makes the classes easier on the eyes. We'd lose the performance benefits of `__slots__`, if that matters. Unlike the freezing method used by attrs and Python 3.7 data classes, this will allow re-assignment of initialized attributes.

If the new implementation is preferred across the board, I can replace this PR's commits with the alternate ones.

ken2812221 commented:
Concept ACK. Will try to learn how `__slots__` work.

promag commented:
Concept ACK, +1 `__slots__` for this case.

maflcko reviewed (approved):
I slightly prefer the decorator for its brevity when used and thus easy reusability. We don't care about performance improvements if they are not significant or can not be measured.

Though, if slots are the commonly the preferred thing, I am fine with that.

jnewbery commented:
Concept ACK either approach. I slightly prefer the `__slots__` approach, since it seems more explicit to me and doesn't require us to roll our own attribute-fixing code, but I'm happy with either.

MarcoFalke commented:
Ok, let's keep the slots then.

practicalswift commented:
Agree with @jnewbery – the explicitness of `__slots__` is nice. Also: it is more idiomatic since it works out of the box without extra code.

jimmysong commented on test/functional/test_framework/script.py:
```diff
@@ -361,8 +361,11 @@ def __init__(self, msg, data):
+
     def __init__(self, d=0):
         self.value = d
 
```
Any reason for not doing `__slots__ = ()` for `CScript` below?

JustinTArthur replied:
I avoided it because it used to be [documented as impossible for variable length sequences](https://docs.python.org/2.5/ref/slots.html#l2h-252). Looks like [recent revisions have been updated](https://docs.python.org/3/reference/datamodel.html#notes-on-using-slots) to clarify that it's ok as long as it's an empty `__slots__`.

I'll revise, test-to-confirm, and adjust the commit.

JustinTArthur replied:
Done

jimmysong reviewed (commented):
ACK

Nice use of the slots feature! I would like to see something instructing new devs in the Python testing documentation (functional/README.md) so people don't trip over it.

JustinTArthur commented on test/functional/p2p_segwit.py:
```diff
@@ -769,12 +769,14 @@ def test_p2sh_witness(self):
+        spend_tx.vin[0].scriptSig = CScript([p2wsh_pubkey, b'a'])
         spend_tx.rehash()
-        test_transaction_acceptance(self.nodes[0], self.test_node, spend_tx, with_witness=False, accepted=False)
+        with self.nodes[0].assert_debug_log(expected_msgs=('Not relaying invalid transaction {}'.format(spend_tx.hash), 'mandatory-script-verify-flag-failed')):
```
Reasonable, will adjust commit.

JustinTArthur commented on test/functional/p2p_segwit.py:
```diff
@@ -769,12 +769,14 @@ def test_p2sh_witness(self):
         # segwit activation.  Note that older bitcoind's that are not
         # segwit-aware would also reject this for failing CLEANSTACK.
-        test_transaction_acceptance(self.nodes[0], self.test_node, spend_tx, with_witness=False, accepted=False)
+        with self.nodes[0].assert_debug_log(expected_msgs=(spend_tx.hash, 'was not accepted: non-mandatory-script-verify-flag')):
```
Reasonable, will adjust commit.

JustinTArthur committed:
Document fixed attribute behavior in critical test framework classes.

Per @jimmysong's suggestion in bitcoin/bitcoin#14305. Also corrects
module for network objects and wrappers.

jimmysong reviewed (approved):
Thanks for the changes!

ACK

jnewbery commented:
utACK e46023287689fc8e79b9a82fe1a827d87c769423

MarcoFalke commented:
utACK e460232876
